### PR TITLE
Xdmod k8s deployment

### PR DESCRIPTION
**Pull request by rob-baron**
Need to write up coherent instruction on how it was installed

Need to write some scripts for the init and setup deployments.  And yes, I think I agree with lars that init containers will do.

### CODE REVIEW

File: Dockerfile.centos7:xdmod-base
Comment (larsks) [inline Dockerfile.centos7:xdmod-base:1]: You mentioned that this was copied from the xdmod project, but I'm not sure that we need it. Installing the `xdmod` rpm itself will bring in any necessary dependencies, so the other Dockerfiles should be fine if they were simply based on a recent CentOS image instead.

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:16]: The bulk of this block (setting up an `rpmbuild` directory, installing an `ssh` server, and installing all the `pcp` related packages) is unnecessary. Probably we don't need the `ssh` client, either. It looks like really we just need to install `python` and then continue with the `pip` commands.

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:21]: These three lines could be replaced by:

```
RUN yum -y install https://github.com/ubccr/xdmod/releases/download/v10.0.0/xdmod-10.0.0-1.0.el7.noarch.rpm
```

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:23]: This file doesn't exist in the repository.

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:25]: It looks like you're just deleting `ssl.conf`.

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:29]: I wonder if test data should just be mounted in a runtime rather than baked into the image?

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:34]: It looks like you're doing some end-of-line conversion here. That would make sense if we were pulling a file from some third-party source, but since it lives right here in the repository why not just ensure it's using correct end-of-line conventions in the repository and remove these `sed` statements?

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:32]: These files don't exist in the repository.

File: Dockerfile.xdmod
Comment (larsks) [inline Dockerfile.xdmod:49]: I think it would make more sense to just have:

```
CMD ["/usr/sbin/httpd", "-D", "FOREGROUND"]
```

There's no reason to set the `ENTRYPOINT` here, and it makes it more difficult if someone just wanted to spawn a shell instead of running Apache.

File: Dockerfile.xdmod.init
Comment (larsks) [inline Dockerfile.xdmod.init:1]: This seems to be largely identical to `Dockerfile.xdmod`. What's the difference between the resulting two images?

File: Dockerfile.xdmod.init
Comment (larsks) [inline Dockerfile.xdmod.init:10]: This comment doesn't match version of xdmod actually installed by the Dockerfile.

File: assets/mysql-server.cnf
Comment (larsks) [inline assets/mysql-server.cnf:1]: This is a configuration file for mysql, but we're never running mysql/mariadb inside the xdmod containers. The database service in the k8s configuration in this repository is provided by the upstream mariadb image.

File: assets/vimrc
Comment (larsks) [inline assets/vimrc:1]: I don't think we need to install a vim configuration into our Docker images. In general we shouldn't be editing things inside the image.

File: xdmod-pv.yaml
Comment (larsks) [inline xdmod-pv.yaml:14]: We generally won't use `hostPath` volumes in our OpenShift deployments. They require elevated privileges, and they can't be accessed from different nodes like a regular PVC.

File: Dockerfile.centos7:xdmod-base
Comment (rob-baron) [reply Dockerfile.centos7:xdmod-base:1]: At this stage the xdmod-base loads tooling that is convenient (vi, wget ... ) that is use in building xdmod.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod:49]: done.

File: xdmod-pv.yaml
Comment (rob-baron) [reply xdmod-pv.yaml:14]: This is how I am creating the PV on minikube.  As I don't need to worry about creating PVs on OpenShift.

Would it be better for me to keep the minicab deployment locally, or should it be under source control?

File: Dockerfile.centos7:xdmod-base
Comment (larsks) [reply Dockerfile.centos7:xdmod-base:1]: We're not actually building xdmod, right? We're just consuming the RPM. But if it's convenient to get things going might as well keep it for now; we can also revisit later.

File: Dockerfile.xdmod.init
Comment (larsks) [reply Dockerfile.xdmod.init:1]: I think a single image would be best (and then just override the command as necessary in your deployment manifests).

File: xdmod-pv.yaml
Comment (larsks) [reply xdmod-pv.yaml:14]: I didn't realize you were explicitly targeting minikube right now. If that's the case, I guess leave these here for now and we'll revisit when we get around to deploying in openshift.

Comment (rob-baron) [reply xdmod-pv.yaml:14]: I am planning on both.  minikube is more convenient for development.

I am keeping it separated, that the deployment to openshift under /k8s/..., and the pv definitions specifically for minikube out of that directory for now.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod:21]: done

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod:25]: done

File: Dockerfile.xdmod.init
Comment (rob-baron) [reply Dockerfile.xdmod.init]: I am now creating a single docker file.  In some regards it should be teased apart into 3 different docker files:

1.  to copy the configuration files and source code to their respective directories
2. to run the xdmod-setup script (doesn't really need the web server, as an example)
3. to run xdmid's UI

but that is for a different day

File: Dockerfile.xdmod.init
Comment (rob-baron) [reply Dockerfile.xdmod.init:10]: done

File: assets/mysql-server.cnf
Comment (rob-baron) [reply assets/mysql-server.cnf:1]: The xdmod folks are running mariadb out of the same container that xdmod is running.  The example that I am working from used a separate container to run mariadb.  I went with running mariadb in a separate container as that made more sense, though it is a bit more difficult to setup.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod:23]: done.

File: assets/vimrc
Comment (rob-baron) [reply assets/vimrc:1]: I find it useful for debugging things within the container.  I will not have in the production container - however nothing seems to be dependent on vim, so it is easily removed.

**Comment by rob-baron** [comment]
Since there is quite a bit to review, might as well start.  I've addressed most of Larsks' original comments, though I plan on doing more pruning of the docker build file, replacing the route with an ingress - and I am sure I'm forgetting some other things to try.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod]: I have eliminate pcp.  However, xdmod seems to require the openssl-server and openssl-clients with the way that they are doing things.  I have simplified this a bit.

File: Dockerfile.xdmod.init
Comment (rob-baron) [reply Dockerfile.xdmod.init]: I eliminated the Dockerfile.xdmod.init.

File: assets/vimrc
Comment (rob-baron) [reply assets/vimrc:1]: Ok, I have removed vi from the "production" container and moved it to a "dev" container based off of the production container.

the build configs/image streams have yaml for them.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod]: I'll move it to the "dev" build.

File: assets/mysql-server.cnf
Comment (rob-baron) [reply assets/mysql-server.cnf:1]: Actually assets/mysql-server.cnf and the vi configuration files are not being used, though they may be useful to include in the development container.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod]: I just kept this as it was the easiest way to get something working.  will look at removing this in the future.

File: Dockerfile.centos7:xdmod-base
Comment (rob-baron) [reply Dockerfile.centos7:xdmod-base]: The base RPM doesn't include all of its dependencies.

File: Dockerfile.xdmod
Comment (rob-baron) [reply Dockerfile.xdmod]: Actually, I can eliminate this as I can copy in the test data with oc rsync.

**Comment by joachimweyl** [comment]
I don't see any open conversations about changes that need to be made. What is the next step for this PR?

File: Dockerfile.moc-xdmod
Comment (larsks) [inline Dockerfile.moc-xdmod:94]: It really shouldn't be necessary to open up access like this on so many directories. We ought to be able to solve these with more granular permissions, and possibly the correct use of volumes for things that need to be writable like `/var/log` and `/run` (the latter of which could be a `tmpfs` mount).

**Comment by joachimweyl** [comment]
Rob - "It is basically done, I have some more testing to do today with respect to the generalizations (which I had to rework over the weekend).  Both, with a couple of the dependencies, disappearing, and realizing the way that I had laid it out wasn’t the best way forward."

**Comment by joachimweyl** [comment]
@rob-baron it appears as though all comments are resolved or almost there. What are the next steps?

File: Dockerfile.moc-xdmod
Comment (rob-baron) [reply Dockerfile.moc-xdmod]: Multiple processes need to read and write to the /root directory as that is where I am mounting xdmod data directory

/run/httpd permissions were too restrictive for it to be started by the container

/etc/xdmod needs to be modified by xdmod-setup 

/var/lib needs read write permissions for xdmod's php script stores information there (I seem to remember a pid file and/or cookies).

These are the directories that xdmod/php uses by default, I cannot recommend we put them all under /var/log.  Furthermore, I cannot recommend mounting volumes at run time as we don't need to persist that state information between runs.

File: README.md
Comment (rob-baron) [reply README.md:104]: I'll change the instruction.

**Comment by rob-baron** [comment]
kim,

My next steps are:

1) finish addressing the comments.
2) get the install on cop-staging working.
3) redo the install on the nerc

File: hypervisor_facts.py
Comment (rob-baron) [reply hypervisor_facts.py:92]: done.

**Comment by joachimweyl** [comment]
@rob-baron  based on your standup notes it sounds like 2 and 3 are done and if your flurry of activity this morning is any indication you are well on your way to addressing all of the comments. Well done.

File: Dockerfile.moc-xdmod
Comment (rob-baron) [reply Dockerfile.moc-xdmod:94]: I was able to change the /run/httpd to just that directory.

Keep in mind, we are running this as a user process and not as root (as is expected by the xdmod team).   So here is a run down of the all of the permissions that you are mentioning:

        chmod -R 777 /usr/bin      # stores the xdmod executables (xdmod-setup, xdmid-shredder ... )
        chmod 777 /usr/bin         # This gave the error that xdmod-setup could not be run

        chmod -R 777 /var/log      # this needs to store log files written by a user process of apache, php and xdmod
        chmod -R 777 /var/lib      # this is used to store the session cookies used by xdmod

        chmod -R 777 /root         # This stores the data
        default                    # unable to create the subdirectories for the data        

        chmod 777 /run/httpd       # This is the directory of the pid file (removed the -R) 

        chmod -R 777 /etc/xdmod     # stores the configuration of xdmod
        default                     #  Was able to copy, but not able to modify  

        chmod -R 777 /etc/httpd  # stores all of the configuration for the apache    
        default                     # Was able to copy, but not able to modify

What permissions would be restrictive enough for you?  Are there a particular set you want me to test?

File: Dockerfile.moc-xdmod
Comment (rob-baron) [reply Dockerfile.moc-xdmod]: I'm marking this as resolved as it is it duplicated below.

File: Dockerfile.moc-xdmod
Comment (rob-baron) [reply Dockerfile.moc-xdmod:94]: @larsks - do you have an opinion on this?

Comment (larsks) [reply Dockerfile.moc-xdmod:94]: Yes, and sorry for the slow reply.

Looking at the deployment, we're mounting the following volumes:

```
- name: vol-xdmod-conf
  mountPath: /etc/xdmod
- name: vol-xdmod-src
  mountPath: /usr/share/xdmod
- name: vol-xdmod-data
  mountPath: /root/xdmod_data
- name: vol-httpd-conf
  mountPath: /etc/httpd
- name: vol-clouds-yaml
  mountPath: /etc/openstack
- name: vol-httpd-conf
  mountPath: /root/httpd
```

The user under which the container is running will already have
write access to all of these directories, although you will not be
able to access directories under `root`. We can demonstrate both facts
with the following example:

```
apiVersion: apps/v1

kind: Deployment
metadata:
  name: permission-test
spec:
  template:
    spec:
      containers:
        - image: docker.io/alpine:latest
          name: permission-test
          volumeMounts:
            - name: vol-xdmod-conf
              mountPath: /etc/xdmod
            - name: vol-xdmod-src
              mountPath: /usr/share/xdmod
            - name: vol-xdmod-data
              mountPath: /root/xdmod_data
          command:
            - sh
            - -c
            - |
              set -e

              for dir in /etc/xdmod /usr/share/xdmod /root/xdmod_data; do
                cd $dir
                pwd
                date > testfile
              done

              sleep inf
      volumes:
        - name: vol-xdmod-conf
          persistentVolumeClaim:
            claimName: pvc-xdmod-conf
        - name: vol-xdmod-src
          persistentVolumeClaim:
            claimName: pvc-xdmod-src
        - name: vol-xdmod-data
          persistentVolumeClaim:
            claimName: pvc-xdmod-data
```

The pod from that deployment will output:

```
/etc/xdmod
/usr/share/xdmod
sh: cd: line 3: can't cd to /root/xdmod_data: Permission denied
```

That is, it demonstrably is able to write files to `/etc/xdmod` and
`/usr/share/xdmod` without making any permissions changes. It is
unable to access `/root/xdmod_data` because the permissions on the
`/root` directory are:

```
$ ls -ld /root
drwx------    1 root     root            24 Jun 10 16:04 /root
```

But that doesn't mean we should change the permissions on `/root`:
because we *know* we're going to be running as a non-`root` user, we
should be using directories in a different location (which could be
pretty much anything: `/data`, `/home/xdmod`, whatever you want).

For directories like `/var/log/xdmod` and `/var/lib/<whatever>`, a
common solution would be to mount an `emptyDir` volume at both those
locations.  E.g., if we were to modify the above example `Deployment`,
we would add the following to the `volumes` section:

```
      volumes:
        - name: xdmod-state
          emptyDir: {}
```


And then the following to the `volumeMounts` section in our
containers:

```
          volumeMounts:
            - name: xdmod-state
              mountPath: /var/log/xdmod
            - name: xdmod-state
              mountPath: /var/lib/xdmod
```

With that volume mounted, while we cannot write to `/var/log` or
`/var/lib`:

```
~ $ touch /var/log/testfile
touch: /var/log/testfile: Permission denied
~ $ touch /var/lib/testfile
touch: /var/lib/testfile: Permission denied
```

We can write to `/var/log/xdmod` and `/var/lib/xdmod` without a
problem:

```
~ $ touch /var/log/xdmod/testfile
~ $ touch /var/lib/xdmod/testfile
```

I would not expect anything to be *writing* to `/usr/bin` at runtime,
so I'm not sure what's going on there.

Comment (larsks) [reply Dockerfile.moc-xdmod:94]: ...and since you're also mounting a volume on `/etc/httpd`, that directory should also be writable by the current user without any permission changes.

Comment (rob-baron) [reply Dockerfile.moc-xdmod:94]: Do we really need a mount a separate disk to hold the pid file as we already have ephemeral storage (as in the container once it is deployed)?

**Comment by joachimweyl** [comment]
@rob-baron it looks like the comments have been almost all dealt with. Do we have any last pieces that we need to change?

**Comment by rob-baron** [comment]
larsks, I'll time box this effort and merge this PR this evening, either with changes that support what you want or without and add an issue.

**Comment by joachimweyl** [comment]
> larsks, I'll time box this effort and merge this PR this evening, either with changes that support what you want or without and add an issue.

@rob-baron just to confirm when you said you would merge the PR with or without the changes. I assume by that you mean if there are no changes you will merge and if there are changes, the changes will require review and therefore it will not be merged yet?

**Comment by rob-baron** [comment]
ok, I'll merge it at the last review state.

**Review by knikolla — APPROVED** [review]

**Pull request merged**
